src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -115,26 +115,31 @@ public User getUser(String email) {
   }
   
   /**
-  * Returns a set of all users who have posted a message
+  * Returns a set of all users who have posted a message or about me
   */
```

Comment:
For now I just kept the return type of method Set<String> instead of Set<User> mainly because I do not how to extract data from the user object in the front-end side of application: conversions to Gson, Json, and getting information from the final Json object if User object is used instead of String. 
user link  is same as user email, so property "user" in Message = property "email" in User
if a user has posted "aboutMe" , the text in "aboutMe" will be concatenated to user link separated by " " space because the user email will not have a white-space in it.
The order of Query("User") and Query("Message") can be different, but the code will be slightly different to make sure all users with Message who have posted "about me", will get their "about me" displayed in the community page while avoiding duplicates. The reason for this difference is that "Message" does not have the property "about me", so if a user with message is added to the HashSet, when the same user is queried in Query("User"), we have to delete the old record and add the new record with about me concatenated to it. To avoid deleting and re-adding, I put Query("User") first and then Query("Message") second.  
Another HashSet<String>, named usersWithAboutMe, is also added to keep track of added records using only user-link(=email) because when adding to user HashSet, after concatenating aboutMe to user-link, we cannot define duplicates of records by user-Link, so we need a different set to keep track of records without concatenation.

Reply:
Alright - I would suggest maybe using '|'/pipe character instead of whitespace, as whitespace has some ambiguity in web: https://en.wikipedia.org/wiki/Whitespace_character#Definition_and_ambiguity

Also, have you considered using an HashMap instead, keyed by the email property? It might save having to use two HashSets.

We can do a follow-up pull request to change it from later. I would recommend walking up from user_page_loader.js#buildMessageDiv (maybe starting from where message.user is called) into DataStore#getMessages (or we can do it together) to figure out how the data is transferred.

Reply:
Wow! these are very helpful and great ideas! I will definitely use '|', and also HashMap (it is exactly what I need!)